frame/base/bli_riscv_detect_arch.h

```diff
@@ -0,0 +1,148 @@
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    - Neither the name(s) of the copyright holder(s) nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+*/
+
+/* Construct a RISC-V architecture string based on available features. */
+
+#if __riscv
+
+#if __riscv_arch_test
+
+#if __riscv_i
+#define RISCV_I i
+#else
+#define RISCV_I
+#endif
+
+#if __riscv_e
+#define RISCV_E e
+#else
+#define RISCV_E
+#endif
+
+#if __riscv_m
+#define RISCV_M m
+#else
+#define RISCV_M
+#endif
+
+#if __riscv_a
+#define RISCV_A a
+#else
+#define RISCV_A
+#endif
+
+#if __riscv_f
+#define RISCV_F f
+#else
+#define RISCV_F
+#endif
+
+#if __riscv_d
+#define RISCV_D d
+#else
+#define RISCV_D
+#endif
+
+#if __riscv_c
+#define RISCV_C c
+#else
+#define RISCV_C
+#endif
+
+#if __riscv_p
+#define RISCV_P p
+#else
+#define RISCV_P
+#endif
+
+#if __riscv_v
+#define RISCV_V v
+#else
+#define RISCV_V
+#endif
+
+#else /* __riscv_arch_test */
+
+/* We assume I and E are exclusive when __riscv_arch_test isn't defined */
+#if __riscv_32e
+#define RISCV_I
+#define RISCV_E e
+#else
+#define RISCV_I i
+#define RISCV_E
+#endif
+
+#if __riscv_flen >= 32
+#define RISCV_F f
+#else
+#define RISCV_F
+#endif
+
+#if __riscv_flen >= 64
+#define RISCV_D d
+#else
+#define RISCV_D
+#endif
+
+/* Cannot test for M, A, C, P if __riscv_arch_test isn't defined */
+#define RISCV_M
+#define RISCV_A
+#define RISCV_C
+#define RISCV_P
+
+#if __riscv_vector
+#define RISCV_V v
+#else
+#define RISCV_V
+#endif
+
+#endif /* __riscv_arch_test */
+
+#define CAT2(a,b) a##b
+#define CAT(a,b) CAT2(a,b)
+
+#if __riscv_arch_test && ( __riscv_i && __riscv_m && __riscv_a && \
+						   __riscv_f && __riscv_d && !__riscv_e)
+
+CAT(rv, CAT(__riscv_xlen, CAT(g, CAT(RISCV_C, CAT(RISCV_P, RISCV_V)))))
+
+#else
+
+CAT(rv, CAT(__riscv_xlen, CAT(RISCV_I, CAT(RISCV_E, CAT(RISCV_M, CAT(RISCV_A,
+CAT(RISCV_F, CAT(RISCV_D, CAT(RISCV_C, CAT(RISCV_P, RISCV_V))))))))))
+
```

Comment:
Have you been able to test the `else` branch? I recall haven gotten a compiler error at some point, complaining that the extensions have to be listed in a certain order. In particular, the order of `p` and `v` is not what I would have expected. I assume that `p` requires the presence of `v` - then the logical order would be `vp`, but I do not know what is correct. Note that this comment also applies to line 139.

Reply:
`P` is listed before `V`, and does not imply `V`. `P` supports DSP SIMD extensions, and is unrelated to `V` vector extensions.

In Table 27.1, page 152, of the [Unprivileged ISA](https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMAFDQC/riscv-spec-20191213.pdf), it lists `P` before `V`. It also lists `__riscv_p` before `__riscv_v` in the [C API spec](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#user-content-architecture-extension-test-macro).

> Table 27.1: Standard ISA extension names. The table also defines the canonical order in which
> extension names must appear in the name string, with top-to-bottom in table indicating first-to-last
> in the name string, e.g., RV32IMACV is legal, whereas RV32IMAVC is not.

`rv64imafdpv` is legal, while `rv64imafdvp` is not.